docs/idm-spec.md

```diff
@@ -0,0 +1,177 @@
+# Identity Manager Specification
+
+This document defines the IDM components, interfaces, protocols and messages.
+Note that this is still a **draft**, meaning it's incomplete and it's likely to be changed many times.
+
+Index:
+
+- [IDM Client](#idm-client)
+- [IDM Wallet](#idm-wallet)
+- [IDM Bridge](#idm-bridge)
+- [Data Types](#data-types)
+
+
+## IDM Client
+
+```js
+// authenticate, unauthenticate & obtain current session
+.authenticate({ Array<CredentialScope> scopes = [] } ?options): Promise<Session>
+.unauthenticate(): Promise
+.getSession(): Promise<Session>
+.onSessionChange((Session session) => {}): Function (to remove the listener)
+
+// signing and verifying
+.sign(ArrayBuffer data, { IdentifiedSignatureKeyType keyType = session, String ?previewUrl } ?options): Promise<IdentifiedSignature>
+.verifySignature(ArrayBuffer data, IdentifiedSignature signature): Promise<Boolean>
+```
+
+TODO: add didAuth
+
+## IDM Wallet
+
+Main scopes:
+
+```js
+.locker
+.storage
+.did
+.identities
+.session
+```
+
+### .locker
+
+```js
+.locker.unlock(LockerType type, (Any challenge) => Any solution): Promise<PrivateKey>
+.locker.lock()
+.locker.isLocked(): Boolean
+.locker.onLockedChange((Boolean locked) => {}): Function (to remove the listener)
+.locker.secret.getPublicKey(): Promise<PublicKey>
+.locker.secret.getPrivateKey(): Promise<PrivateKey>
+.locker.idle.setMaxTime(Number idleTime)
+.locker.idle.getMaxTime(): Number
+.locker.idle.getRemainingTime(): Number
+.locker.idle.reset():
+.locker.locks.list(): Promise<Array<LockerType>>
+.locker.locks.set(LockerType type, Any solutions): Promise
+.locker.locks.drySet(LockerType type, Any solutions): Promise
+.locker.locks.unset(LockerType type): Promise
+```
+
+### .storage
+
+```js
+.storage.get(String<String>|String key): Promise<Object<String,Any>>
```

Comment:
Is `String<String>` meant to be `Array<String>`?

Reply:
Fixed